src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,272 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        private Node(T element) {
+            this.item = element;
+        }
+    }
+
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public MyLinkedList() {
+        size = 0;
+        head = null;
+        tail = null;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> elementToAdd = new Node<>(value);
+
+        if (size == 0) {
+            head = elementToAdd;
+            tail = elementToAdd;
+        } else {
+            tail.next = elementToAdd;
+            elementToAdd.prev = tail;
+            tail = elementToAdd;
+        }
+
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " does not exist");
+        }
+
+        Node<T> elementToAdd = new Node<>(value);
+
+        if (size == 0) {
+            head = elementToAdd;
+            tail = elementToAdd;
+        } else if (index == size - 1) {
+            Node<T> elPrevTail = tail.prev;
+            elPrevTail.next = elementToAdd;
+            elementToAdd.prev = elPrevTail;
+
+            tail.prev = elementToAdd;
+            elementToAdd.next = tail;
+        } else if (size == index) {
+            tail.next = elementToAdd;
+            elementToAdd.prev = tail;
+            tail = elementToAdd;
+        } else if (index == 0 && size != 0) {
+            Node<T> currentElement = head;
+            head = elementToAdd;
+            elementToAdd.next = currentElement;
+            currentElement.prev = elementToAdd;
+        } else {
+            int countOfIndex = 0;
+            Node<T> element = head;
+
+            while (countOfIndex != index) {
+                element = element.next;
+                countOfIndex++;
+            }
+
+            Node<T> elPrevElement = element.prev;
+            elPrevElement.next = elementToAdd;
+            elementToAdd.prev = elPrevElement;
+
+            elementToAdd.next = element;
+            element.prev = elementToAdd;
+        }
+
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (!list.isEmpty()) {
```

Comment:
In the `addAll` method, check if the current list is empty before accessing `tail` to avoid a `NullPointerException`.